boringtun/src/device/mod.rs

```diff
@@ -621,16 +634,41 @@ impl Device {
 
                     let peer = match &parsed_packet {
                         Packet::HandshakeInit(p) => {
-                            parse_handshake_anon(private_key, public_key, p)
+                            parse_handshake_anon(&private_key, &public_key, p)
                                 .ok()
                                 .and_then(|hh| {
-                                    d.peers
-                                        .get(&x25519_dalek::PublicKey::from(hh.peer_static_public))
+                                    let pub_key =
+                                        x25519_dalek::PublicKey::from(hh.peer_static_public);
+
+                                    d.try_writeable(
+                                        |d| d.trigger_yield(),
+                                        |d| {
+                                            d.cancel_yield();
+                                            match d.registry.get(&pub_key) {
+                                                RegistryPeer::Peer(peer) => Some(peer),
+                                                RegistryPeer::Candidate(peer_candidate) => {
+                                                    d.update_peer(
+                                                        peer_candidate.public_key.clone(),
+                                                        false,
+                                                        false,
+                                                        None,
+                                                        peer_candidate.allowed_ips.as_slice(),
+                                                        peer_candidate.keepalive,
+                                                        None,
+                                                    );
+
+                                                    d.registry.get(&pub_key).into()
```

Comment:
Should this be a call to put the Peer? Maybe I'm missing something, but it feels like at this point we should be promoting this into a full Peer.

Reply:
You've hit on one of the warts with the API as it currently exists. `update_peer` not only creates the full `peer` but also calls `registry.insert`.  I could possibly refactor that portion and require that the callers explicitly add the peer but it would entail adding another method to the `Device` struct as the api would be one of the places to update.  Not a big issue but not sure it's worth the effort with the expected API changes coming from upstream.

Reply:
Not sure if that's a wart 🤔  Just an unfamiliar way of doing things